src/utils/merge-nodes-with-matching-flavors.ts

```diff
@@ -0,0 +1,166 @@
+    importFlavorType,
+    mergeableImportFlavors,
+} from '../constants';
+import type { MergeNodesWithMatchingImportFlavors } from '../types';
+import { getImportFlavorOfNode } from './get-import-flavor-of-node';
+
+type MergeableFlavor = typeof mergeableImportFlavors[number];
+function isMergeableFlavor(flavor: string): flavor is MergeableFlavor {
+    return mergeableImportFlavors.includes(flavor as MergeableFlavor);
+}
+
+function selectMergeableNodesByImportFlavor(
+    nodes: ImportDeclaration[],
+): Record<MergeableFlavor, ImportDeclaration[]> {
+    return nodes.reduce(
+        (groups, node) => {
+            const flavor = getImportFlavorOfNode(node);
+            if (isMergeableFlavor(flavor)) {
+                groups[flavor].push(node);
+            }
+            return groups;
+        },
+        {
+            [importFlavorRegular]: [] as ImportDeclaration[],
+            [importFlavorType]: [] as ImportDeclaration[],
+        },
+    );
+}
+
+function selectNodeImportSource(node: ImportDeclaration) {
+    return node.source.value;
+}
+
+function nodeIsImportNamespaceSpecifier(
+    node: ImportSpecifier | ImportDefaultSpecifier | ImportNamespaceSpecifier,
+): node is ImportNamespaceSpecifier {
+    return node.type === 'ImportNamespaceSpecifier';
+}
+function nodeIsImportDefaultSpecifier(
+    node: ImportSpecifier | ImportDefaultSpecifier | ImportNamespaceSpecifier,
+): node is ImportDefaultSpecifier {
+    return node.type === 'ImportDefaultSpecifier';
+}
+
+/** Return false if the merge will produce an invalid result */
+function mergeIsSafe(
+    nodeToKeep: ImportDeclaration,
+    nodeToForget: ImportDeclaration,
+) {
+    if (
+        nodeToKeep.specifiers.some(nodeIsImportNamespaceSpecifier) ||
+        nodeToForget.specifiers.some(nodeIsImportNamespaceSpecifier)
+    ) {
+        // An `import * as Foo` namespace specifier cannot be merged
+        //   with other import expressions.
+        return false;
+    }
+    if (
+        nodeToKeep.specifiers.some(nodeIsImportDefaultSpecifier) &&
+        nodeToForget.specifiers.some(nodeIsImportDefaultSpecifier)
+    ) {
+        // Two `import Foo from` specifiers cannot be merged trivially.
+        // -- Notice: this is *not* import {default as Foo1, default as Foo2} -- that's legal!
+        //
+        // Future work could convert `import Foo1 from 'a'; import Foo2 from 'a';
+        //  into `import {default as Foo1, default as Foo2} from 'a';`
+        // But since this runs the risk of making code longer, this won't be in v1.
+        return false;
+    }
+    return true;
+}
+
+/**
+ * @returns (true to delete node | false to keep node)
+ */
+function mergeNodes(
+    nodeToKeep: ImportDeclaration,
+    nodeToForget: ImportDeclaration,
+) {
+    if (!mergeIsSafe(nodeToKeep, nodeToForget)) {
+        return false;
+    }
+
+    nodeToKeep.specifiers.push(...nodeToForget.specifiers);
+
+    // The line numbers will be all messed up. Is this a problem?
```

Comment:
[There's the part](https://github.com/IanVS/prettier-plugin-sort-imports/commit/b5c23c79174682452ccd46ec2feddc49448880a5#diff-ce50d77c8138cb15e5b03ef96053bc615ad9450e5c30ada045bf72de79d6966eR164-R195) where it removes nodes from the original code which relies on the line number being correct (it did so even before the linked commit). 

It is called from https://github.com/IanVS/prettier-plugin-sort-imports/blob/b5c23c79174682452ccd46ec2feddc49448880a5/src/utils/get-code-from-ast.ts#L35

I think that might be alright since it should operate on the original nodes, but perhaps you could check again that the nodes passed to that function are not the modified ones?

Reply:
The nodes passed do both the modified & the original ones -- exactly line: https://github.com/IanVS/prettier-plugin-sort-imports/pull/19/commits/cf4e6af3093740dd1b035c3d798ae8ddb7af10ca#diff-ab5d98b73e57fb0fe6cb1b2b9ea7d14530182d5d1987eae7c545bd15f4528cfeR37

The trick is that the modified nodes don't have their line-numbers adjusted. The line-numbers still match what's in the provided code.

So… I think we're good?

Reply:
Yes, I think so. If not, and if it did remove code at the wrong position, then the tests would have failed already

Reply:
I don't quite understand this, and it relates to my comment from earlier about why we're providing both the modified and original nodes.  But, feel free to ignore me, if y'all think it's fine how it is.

Reply:
That comment was a bit out of date, as I've now learned that that's actually an important attribute of how `getCodeFromAst` functions. It needs the old line-numbers/loc to properly move things around!

```ts
// These mutations don't update the line numbers, and that's crucial for moving things around.
// To get updated line-numbers you would need to re-parse the code after these changes are rendered!
```